core/internal/runhistoryreader/parquet/parition_reader.go

```diff
@@ -0,0 +1,184 @@
+			}
+		}
+	}
+}
+
+func (p *ParquetPartitionReader) iterator() (RowIterator, error) {
+	select {
+	case <-p.ctxDone:
+		return nil, p.ctx.Err()
+	case <-p.dataReady:
+		p.executor.Wait()
+		return p.it, p.err
+	}
+}
+
+// NewParquetPartitionReader creates a new parquet partition reader for the given file.
+// keys argument is the list of columns to include in the row iterator.
+// opt argument is used to filter out rows based on a key and range of values.
+func NewParquetPartitionReader(
+	ctx context.Context,
+	file *pqarrow.FileReader,
+	keys []string,
+	opt RowIteratorOption,
+) *ParquetPartitionReader {
+	makeIterator := func(partition *ParquetPartitionReader) (RowIterator, error) {
+		return NewRowIterator(ctx, file, keys, opt)
+	}
+
+	return newParquetPartitionReader(ctx, file, makeIterator)
+}
+
+func newParquetPartitionReader(
+	ctx context.Context,
+	file *pqarrow.FileReader,
+	makeIterator func(*ParquetPartitionReader) (RowIterator, error),
+) *ParquetPartitionReader {
+	partition := &ParquetPartitionReader{
+		ctx:       ctx,
+		ctxDone:   ctx.Done(),
+		executor:  parallel.Unlimited(ctx),
+		file:      file,
+		dataReady: make(chan struct{}),
+	}
+
+	partition.executor.Go(func(ctx context.Context) {
+		defer close(partition.dataReady)
+		partition.it, partition.err = makeIterator(partition)
+	})
+
+	return partition
+}
+
+// GetColumnRangeFromStats returns the range of values
+// for the given column index in the given row group.
+func GetColumnRangeFromStats(
+	meta *metadata.FileMetaData,
+	rowGroupIndex int,
+	columnName string,
+) (minValue float64, maxValue float64, err error) {
+	minValue = -1
+	maxValue = -1
+	if rowGroupIndex < 0 || rowGroupIndex >= len(meta.RowGroups) {
+		err = fmt.Errorf("%s row group not found in parquet file", columnName)
+		return
+	}
+
+	// *parquet.RowGroup is an internal package to arrow, so we cannot reference it directly...
+	rowGroup := meta.RowGroups[rowGroupIndex]
+	colIndex := meta.Schema.ColumnIndexByName(columnName)
+	if colIndex < 0 || colIndex >= len(rowGroup.Columns) {
+		err = fmt.Errorf("%s column not found in parquet file", columnName)
+		return
+	}
+
+	statistics := rowGroup.Columns[colIndex].MetaData.Statistics
+	if statistics == nil {
+		err = fmt.Errorf("missing statistics for %s column chunk", columnName)
+		return
+	}
+
+	switch t := meta.Schema.Column(colIndex).PhysicalType(); t {
+	case parquet.Types.Int64:
+		imin, err := decodeValue[int64](statistics.MinValue)
+		if err != nil {
+			return -1, -1, fmt.Errorf("%w", err)
+		}
+		imax, err := decodeValue[int64](statistics.MaxValue)
+		if err != nil {
+			return -1, -1, fmt.Errorf("%w", err)
+		}
+		minValue, maxValue = float64(imin), float64(imax)
+	case parquet.Types.Double:
+		minValue, err = decodeValue[float64](statistics.MinValue)
+		if err != nil {
+			return -1, -1, fmt.Errorf("%w", err)
+		}
+		maxValue, err = decodeValue[float64](statistics.MaxValue)
+		if err != nil {
+			return -1, -1, fmt.Errorf("%w", err)
```

Comment:
These all wrap the error (`%w`) but the function doesn't document what errors can be expected, so callers wouldn't be able to unwrap errors anyway. Should this use `%v` instead?